src/Calypso-SystemPlugins-SUnit-Browser/TClyGenerateTestClass.trait.st

```diff
@@ -0,0 +1,73 @@
+"
+Description
+--------------------
+
+I am a trait containing the behavior or retrieving the test class of a class.
+
+If the test class does not exists I am able to generate it.
+"
+Trait {
+	#name : #TClyGenerateTestClass,
+	#category : #'Calypso-SystemPlugins-SUnit-Browser-Traits'
+}
+
+{ #category : #action }
+TClyGenerateTestClass >> addNewCommentForTestClass: aClass basedOn: baseClass [
+	aClass
+		comment:
+			(String
+				streamContents: [ :stream | 
+					| name |
+					name := aClass name.
+					name first isVowel
+						ifTrue: [ stream nextPutAll: 'An ' ]
+						ifFalse: [ stream nextPutAll: 'A ' ].
+					stream
+						nextPutAll: name;
+						nextPutAll: ' is a test class for testing the behavior of ';
+						nextPutAll: baseClass name ])
+]
+
+{ #category : #testing }
+TClyGenerateTestClass >> isValidClass: inputClass [
+	^ (inputClass isTestCase or: [ inputClass isMeta ]) not
+]
+
+{ #category : #action }
+TClyGenerateTestClass >> newTestClassCategoryFor: aClass [
+	| tag |
+	tag := aClass package classTagForClass: aClass.
+	^ String
+		streamContents: [ :s | 
+			s
+				nextPutAll: aClass package name;
+				nextPutAll: '-Tests'.
+			tag isRoot
+				ifFalse: [ s
+						nextPut: $-;
+						nextPutAll: tag name ] ]
+]
+
+{ #category : #accessing }
+TClyGenerateTestClass >> testClassFor: inputClass [
+	| className resultClass |
+	className := self testClassNameFor: inputClass.
+	self class environment
+		at: className
+		ifPresent: [ :class | resultClass := class ]
+		ifAbsent: [
+			(self isValidClass: inputClass) ifFalse: [ ^ nil ].
+			RPackageOrganizer default ensureExistAndRegisterPackageNamed: inputClass package name asString , '-Tests'.
+			resultClass := TestCase
+				subclass: className
+				instanceVariableNames: ''
+				classVariableNames: ''
+				package: (self newTestClassCategoryFor: inputClass).
```

Comment:
But more important problem here are global RPackageOrganizer and "self class environment".
In Ring scenario or remote scenario all these commands should continue work. It means that actual environment or package organizer can't be hardcoded and should be parameters of commands.
So user should be able specify them and they should be available from execution context.

In calypso you will need only ClySystemEnvironment instance under browser:
```Smalltalk
AGenerationCommand>>prepareFullExecutionInContext: aContext
   super ...
   ...
   systemEnvironment := aContext systemEnvironment.
```
ClySystemEnvironment includes methods to create packages and classes. Look at SycAddNewPackageCommand for example